job wrapper - python3, new script and cleanEnv=False

mapellidario opened the pull request:
Related to #7084 

### status

- [x] tested in test11
  - task [221005_154507:dmapelli_crab_20221005_174504](https://cmsweb-test11.cern.ch/crabserver/ui/task/221005_154507%3Admapelli_crab_20221005_174504) on test11 / vocms059
- [ ] tested in testbed

### overview

This is the first PR related to the modernization of CRAB's job wrapper.

It achieves three different goals, each in its own commit

- modernize the scripts, so that they all works with python3. This requires https://github.com/dmwm/WMCore/pull/11321, which will be available with WMCore `2.1.4`, or we can get a preview with `2.1.4.pre4`
- add a new script that is intended to be shared between WMCore and CRAB. For the time being, i added it to CRABServer repo, then we will think about moving it to WMCore after having a chat with them
- when using WMCore's `Scram.py:Scram.__call__()`, use `cleanEnv=False`

### notable changes

Since we do not clean the environment anymore, instead of setting env variables in `CMSRunAnalysis.py`, we should set them in the function `env_set_local` in `env-CMSRunAnalysis.sh`, so that they can be shared with WMCore.

I changed how we save the startup environment, so that bash functions are properly exported

```diff
- set > startup_environment.sh
- sed -e 's/^/export /' startup_environment.sh > tmp_env.sh
- mv tmp_env.sh startup_environment.sh
+ declare -p > startup_environment.sh
```
### what to look for in the tests

- on T1_UK_RAL, `XrdSecGSISRVNAMES` should flow automatically from the node env to the cmsRun env
- `SITECONFIG_PATH`, should flow automatically from CMSRunAnalysis.sh to the cmsRun env, simply calling `env_cms_load` from `env-CMSRunAnalysis.sh`. this is necessary for CMSSW 12_6_X

belforte commented:
about **what to look for** above: why not also $X509_USER_PROXY ?

belforte commented on scripts/CMSRunAnalysis.py:
```diff
@@ -634,15 +634,6 @@ def StripReport(report):
-        # no matter what we run, it is very likely to need proxy location
-        preCmd = 'export X509_USER_PROXY=%s; ' % os.getenv('X509_USER_PROXY')
-        # needed for root problem with $HOME/.root.mimes, #6801
-        preCmd += 'export HOME=${HOME:-$PWD}; '
```
where is `$HOME` defined then ?

belforte replied:
OK. We have it in the following script  `CMSRunAnalysis.sh

belforte commented on scripts/env-CMSRunAnalysis.sh:
```diff
@@ -0,0 +1,136 @@
+    echo -e "========  CMS environment load finished at $(TZ=GMT date) ========\n"
+}
+
+load_comp_python() {
```
why not `env_python_load` for consistency with `env_cms_load` above ?
BTW I'd rather put verb first in those names, e.g. : `save_env`, `setup_cms_env`, `setup_python`, `set_env_local`

mapellidario replied:
I always struggle with assigning names, thanks for pointing this out. I propose:

- `save_env`
- `setup_local_env` or `setup_env_local`, i can not decide what i prefer
- `setup_cmsset` (i realized I do not like `setup_cms_env`, too close to `cmsenv` which does something different from `cmsset_default.sh`)
- `setup_python_comp`

belforte commented on scripts/gWMS-CMSRunAnalysis.sh:
```diff
@@ -6,15 +6,16 @@
-export JOBSTARTDIR=$PWD
+# import some auxiliary functions from a script that is intented to be shared
+# with WMCore
+source ./env-CMSRunAnalysis.sh
```
this line should go after line 13 (Startup environment)

mapellidario replied:
eventually, I agree. I think i moved it three times myself. I will fix it

belforte commented on scripts/gWMS-CMSRunAnalysis.sh:
```diff
@@ -87,11 +88,9 @@ echo "Hostname:   $(hostname -f)"
 echo "System:     $(uname -a)"
 echo "Arguments are $@"
 
+# dario: what does this do? do we need this?
```
this line makes stderr to be redirected to stdout for all commands executed in the following part of the script. See e.g. https://stackoverflow.com/a/13088401
Whether is needed, I do not know on the spot.

mapellidario replied:
After thinking about this a bit more, I think we want this line because of how condor deals with command outputs. In DagmanCreator we have

```plaintext
Executable = gWMS-CMSRunAnalysis.sh
Output = job_out.$(CRAB_Id)
Error = job_err.$(CRAB_Id)
```

but in all CRABServer we have some logic about how to deal with `job_out.*`, not `job_err.*`. I think this `exec 2>&1` makes sure that `job_err.*` is empty.

Since it is pretty handy to have both stdout and stderr in the same file, so I would not change anything here! :)

thanks Stefano!

belforte replied:
I suggest adding a comment and/or a ptr to stackoverflow, or I will be confused again in 2 months !

belforte commented on scripts/gWMS-CMSRunAnalysis.sh:
```diff
@@ -246,7 +175,7 @@ echo "======== python bootstrap for stageout at $(TZ=GMT date) FINISHED ========
+PYTHONUNBUFFERED=1 $pythonCommand cmscp.py
 STAGEOUT_EXIT_STATUS=$?
 
 if [ ! -e wmcore_initialized ];
```
is this a good time to review what this thing really is and whether we need it ? There should be a better way to flag that imports from WMCore fail in our python scripts ! Also: message in Line 183 can be made more clear

mapellidario replied:
addressed in this comment: https://github.com/dmwm/CRABServer/pull/7441#issuecomment-1326764235

belforte commented:
Overall looks good and I like how things are organized. I have made some small comments on specific points.
There are two important things though:
 - [ ] the script to be shared with WMCore should not be called `env-CMSRunAnalysis.sh` since it will not be specific to "Anlaysis" jobs
 - [ ] did you test with CMSSW_6 on SL6 containers ?

belforte commented:
gee... I wrote this yesterday...oh well..
Overall looks good and I like how things are organized. I have made some small comments on specific points.
There are two important things though:
 - [ ] the script to be shared with WMCore should not be called `env-CMSRunAnalysis.sh` since it will not be specific to "Anlaysis" jobs
 - [ ] did you test with CMSSW_6 on SL6 containers ?

mapellidario commented:
Stefano, thank you very much for the review!

- I agree that we should rename the script, i just renamed it from `env-CMSRun..` to `submit_env.sh`. We will see if Alan agrees
- I added the tests on CMSSW_6 to the list of tests that I should make before we merge this PR
- I agree that we should review how we flag that we are not WMCore (in particular why we need `wmcore_initialized`). However I do not consider it the scope of this review, but the next one(s). I have added it to mylist here https://github.com/dmwm/CRABServer/issues/7084#issuecomment-1192371547 

At the moment I am focusing on getting reproducible results, using freshly built images, after months of manually tweaking the TW environment inside the container. When adding new files into the mix, half of the problems appear only when you build the rpms!

belforte commented:
> how we flag that we are not WMCore (in particular why we need wmcore_initialized
well. it is much small than that. Current code is simply meant to point out the (in current situation very unrealistic) situation where jobs wrapper starts and runs otherwise, but WMCore was not properly initialized and imports fails. Note that we import WMCore not from CVMFS/..COMP.. but from the tarball sent with the job (to have consistent WMCore with what is used in TW and sched).

mapellidario commented:
We had two open concerns.

First of all, the support of cmssw 7 with slc6: this has been covered by https://github.com/mapellidario/CRABServer/issues/36, so I will mark it as done.

Then you asked if we could do improve checking if WMCore has been imported properly. 

We currently create a file after WMCore has been imported, and we do it twice:

- when `CMSRunAnalysis.sh` calls `CMSRunAnalysis.py`, [1]
- when `gWMS-CMSRunAnalysis.sh` calls `cmscp.py` [2]

I see two ways of achieving this. The first would be to wrap all the imports in a try/except clause, return a specific exit code when the import fails. The second is the one we are currently using, which is creating a file when import is ok, then checking from the bash script if the file is there or not.

The first approach is actually used a well in CMSRunAnalysis.py!

https://github.com/dmwm/CRABServer/blob/82f8d5a6f22a9a38377792e788465da32d5b3fcb/scripts/CMSRunAnalysis.py#L573-L594

Then the natural question is: why do we use both methods (try/except, checking if a file is on disk) in CMSRunAnalysis.py? Having a look at the git history, Brian intentionally added both of them at the same time, https://github.com/dmwm/CRABServer/commit/a3c845ef625cc1b2836e096d9d4288aef2864461. And I think that the answer is in the comment: 

> if we encounter a python segfault, the wrapper will look to see if this file exists and report accordingly

Segmentation faults can not be caught in a try/except, they just kill the python runtime. there is no indication of what happened. So, how do we differentiate from a segmentation fault that happens when we import WMCore from a one happening in cmsRun? We have a redundant method to check if the import is ok.

My question is: did we ever witnessed a segmentation fault while importing from wmcore? I did not, but I guess that I am not paid because of my experience, I trust Brian more than me.

---

[1] 

 https://github.com/dmwm/CRABServer/blob/82f8d5a6f22a9a38377792e788465da32d5b3fcb/scripts/CMSRunAnalysis.sh#L157

https://github.com/dmwm/CRABServer/blob/82f8d5a6f22a9a38377792e788465da32d5b3fcb/scripts/CMSRunAnalysis.py#L593-L594

[2]

https://github.com/dmwm/CRABServer/blob/82f8d5a6f22a9a38377792e788465da32d5b3fcb/scripts/gWMS-CMSRunAnalysis.sh#L252

https://github.com/dmwm/CRABServer/blob/82f8d5a6f22a9a38377792e788465da32d5b3fcb/scripts/cmscp.py#L32-L33

belforte commented:
thanks for the detailed analysis and investigation Dario. I agree that it is wiser to leave things as they are. I have never seen such segmentation faults either, but surely we can't rule out the possibility, maybe is just something from a "how to write super robust code" course in Brian's education, good practice that never hurts to have as habit.

mapellidario commented:
Ok then, I think we can merge this. I have only one question: do we want to print the node / startup environment? 
It has been useful when developing the code, but maybe will only create more confusion for users. However we can also think about keeping it in production for a couple of weeks, then remove after we confirm that everything is ok.

mapellidario commented on scripts/submit_env.sh:
```diff
@@ -0,0 +1,138 @@
+    # dirname -- "$0"
+    # echo $PWD
+    # ls -lrth
+    cat startup_environment.sh
```
my question is: do we want to keep this `cat`?

belforte replied:
no.

belforte replied:
only you can judge if it is worth to have it in production for a week and then remove.

mapellidario replied:
I will remove it, then. 

In case of problems, we can always use a scriptExe that dumps that file. Or we can think about copying it back to the schedd

mapellidario replied:
I removed the `cat` and some other debugging prints. I added to this comment https://github.com/dmwm/CRABServer/issues/7084#issuecomment-1192371547 a note about copying the file `start_environment.sh` back to schedd so that i dont forget about this idea.

I think this PR is now ready.

mapellidario commented:
After confirmation in private chat, I will merge this PR, build new images, deploy them to preprod.
